test/e2e/workspacetype/controller_test.go

```diff
@@ -0,0 +1,217 @@
+				require.Emptyf(t, workspace.Status.Initializers, "workspace has initializers")
+			},
+		},
+		{
+			name: "create a workspace with an explicit non-existing type",
+			work: func(ctx context.Context, t *testing.T, server runningServer) {
+				t.Logf("Create a workspace with explicit non-existing type")
+				workspace, err := server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Create(ctx, &tenancyv1alpha1.ClusterWorkspace{
+					ObjectMeta: metav1.ObjectMeta{Name: "myapp"},
+					Spec:       tenancyv1alpha1.ClusterWorkspaceSpec{Type: "Foo"},
+				}, metav1.CreateOptions{})
+				require.Error(t, err, "failed to create workspace")
+
+				t.Logf("Create type Foo")
+				_, err = server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaceTypes().Create(ctx, &tenancyv1alpha1.ClusterWorkspaceType{
+					ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+					Spec:       tenancyv1alpha1.ClusterWorkspaceTypeSpec{},
+				}, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create workspace type")
+
+				t.Logf("Create workspace with explicit type Foo again")
+				require.Eventually(t, func() bool {
+					workspace, err = server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Create(ctx, &tenancyv1alpha1.ClusterWorkspace{
+						ObjectMeta: metav1.ObjectMeta{Name: "myapp"},
+						Spec:       tenancyv1alpha1.ClusterWorkspaceSpec{Type: "Foo"},
+					}, metav1.CreateOptions{})
+					return err == nil
+				}, wait.ForeverTestTimeout, time.Millisecond*100, "failed to create workspace even with type")
+				require.Equal(t, workspace.Spec.Type, "Foo")
+
+				t.Logf("Expect workspace to become ready because there are no initializers")
+				err = server.orgExpect(workspace, ready)
+				require.NoError(t, err)
+			},
+		},
+		{
+			name: "create a workspace with a type that has initializers",
+			work: func(ctx context.Context, t *testing.T, server runningServer) {
+				t.Logf("Create type Foo with initializers")
+				_, err := server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaceTypes().Create(ctx, &tenancyv1alpha1.ClusterWorkspaceType{
+					ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+					Spec: tenancyv1alpha1.ClusterWorkspaceTypeSpec{
+						Initializers: []tenancyv1alpha1.ClusterWorkspaceInitializer{"a", "b"},
+					},
+				}, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create workspace type")
+
+				t.Logf("Create workspace with explicit type Foo again")
+				var workspace *tenancyv1alpha1.ClusterWorkspace
+				require.Eventually(t, func() bool {
+					workspace, err = server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Create(ctx, &tenancyv1alpha1.ClusterWorkspace{
+						ObjectMeta: metav1.ObjectMeta{Name: "myapp"},
+						Spec:       tenancyv1alpha1.ClusterWorkspaceSpec{Type: "Foo"},
+					}, metav1.CreateOptions{})
+					return err == nil
+				}, wait.ForeverTestTimeout, time.Millisecond*100, "failed to create workspace even with type")
+
+				t.Logf("Expect workspace to be stuck in initializing phase")
```

Comment:
Is this part of the test telling us anything that the unit testing isn't already guaranteeing? i.e. that a workspace is stuck in initialization phase so long as it has initializers.

Unless I'm missing something covering the corner-cases of timing and multiple initializers in e2e is unnecessary. I think it should be sufficient that this test create a workspace with an initializer, remove the initializer, and see the workspace become ready.

Reply:
This here also checks that nobody removes the initializers.

Reply:
Who else would be removing arbitrary initializers?

Reply:
Hopefully nobody :D Workspace controller for example.

Reply:
maybe we add a generic controller for workspaces, bootstrapping something. Then it better only removes its own initializer.

Reply:
Still seems to me like something that could be trivially validated in a unit test: for a given controller's mutation of a workspace, ensure that it does not remove initializers.  But I digress.